### PR TITLE
Added optional `$checkHeader` param to loginRequired for checking the…

### PR DESCRIPTION
**Pull request by SamMousa**
… Accept header; defaults to `true`.

### CODE REVIEW

**Comment by SamMousa** [comment]
#10937

**Comment by SamMousa** [comment]
Can someone look at why this test is failing?
It runs fine locally and the failing test doesn't even have anything to do with my code.

**Comment by SilverFire** [comment]
I've broken it. Fixed in https://github.com/yiisoft/yii2/commit/4dc771d0a0f092dc56d8dd002329fb0ed1022243
Update from master, please

**Comment by SamMousa** [comment]
I think this is ready for merge, let me know if you need anything from me!

**Comment by SilverFire** [comment]
Thank you for the notification. I'll verify it soon

File: framework/web/User.php [hunk 1]
```diff
@@ -136,6 +136,10 @@ class User extends Component
      * @var string the session variable name used to store the value of [[returnUrl]].
      */
     public $returnUrlParam = '__returnUrl';
+    /**
+     * @var array MIME types for which this component should redirect to the loginUrl.
+     */
+    public $acceptableRedirectTypes = ['text/html', 'application/xhtml+xml'];
```
Comment (cebe) [inline framework/web/User.php]: need `@since` annotation.

File: framework/web/User.php [hunk 2]
```diff
@@ -413,16 +417,24 @@ public function setReturnUrl($url)
      *
      * @param boolean $checkAjax whether to check if the request is an AJAX request. When this is true and the request
      * is an AJAX request, the current URL (for AJAX request) will NOT be set as the return URL.
+     * @param boolean $checkAcceptHeader whether to check if the request accepts HTML responses. When this is true and
+     * the request does not accept HTML responses the current URL will not be SET as the return URL. Also instead of
+     * redirecting the user an ForbiddenHttpException is thrown.
```
Comment (cebe) [inline framework/web/User.php]: Add version info:

> This parameter is available since version 2.0.8.

File: framework/web/User.php [hunk 3]
```diff
@@ -669,6 +681,26 @@ public function can($permissionName, $params = [], $allowCaching = true)
     }
 
     /**
+     * Checks if the `Accept` header contains a content type that allows redirection to the login page.
+     * The login page is assumed to serve `text/html` or `application/xhtml+xml`, configure this via
+     * `$acceptableRedirectTypes`
+     * @return boolean whether this request may be redirected to the login page.
```
Comment (cebe) [inline framework/web/User.php]: need `@since` annotation.

File: framework/web/User.php [hunk 4]
```diff
@@ -413,16 +418,24 @@ public function setReturnUrl($url)
...
+     * @param boolean $checkAcceptHeader whether to check if the request accepts HTML responses. When this is true and
+     * the request does not accept HTML responses the current URL will not be SET as the return URL. Also instead of
+     * redirecting the user an ForbiddenHttpException is thrown. This parameter is available since version 2.0.8.
      * @return Response the redirection response if [[loginUrl]] is set
-     * @throws ForbiddenHttpException the "Access Denied" HTTP exception if [[loginUrl]] is not set
+     * @throws ForbiddenHttpException the "Access Denied" HTTP exception if [[loginUrl]] is not set or a redirect is
+     * not applicable (see $checkAcceptHeader).
      */
-    public function loginRequired($checkAjax = true)
+    public function loginRequired($checkAjax = true, $checkAcceptHeader = true)
```
Comment (SilverFire) [inline framework/web/User.php]: Method signature change should be mentioned in UPGRADE.md

File: framework/web/User.php [hunk 5]
```diff
@@ -413,16 +418,24 @@ public function setReturnUrl($url)
      *
      * @param boolean $checkAjax whether to check if the request is an AJAX request. When this is true and the request
      * is an AJAX request, the current URL (for AJAX request) will NOT be set as the return URL.
+     * @param boolean $checkAcceptHeader whether to check if the request accepts HTML responses. When this is true and
+     * the request does not accept HTML responses the current URL will not be SET as the return URL. Also instead of
+     * redirecting the user an ForbiddenHttpException is thrown. This parameter is available since version 2.0.8.
      * @return Response the redirection response if [[loginUrl]] is set
-     * @throws ForbiddenHttpException the "Access Denied" HTTP exception if [[loginUrl]] is not set
+     * @throws ForbiddenHttpException the "Access Denied" HTTP exception if [[loginUrl]] is not set or a redirect is
+     * not applicable (see $checkAcceptHeader).
```
Comment (SilverFire) [inline framework/web/User.php]: `@see checkAcceptHeader`

File: framework/web/User.php [hunk 6]
```diff
@@ -413,16 +418,24 @@ public function setReturnUrl($url)
      *
      * @param boolean $checkAjax whether to check if the request is an AJAX request. When this is true and the request
      * is an AJAX request, the current URL (for AJAX request) will NOT be set as the return URL.
+     * @param boolean $checkAcceptHeader whether to check if the request accepts HTML responses. When this is true and
+     * the request does not accept HTML responses the current URL will not be SET as the return URL. Also instead of
+     * redirecting the user an ForbiddenHttpException is thrown. This parameter is available since version 2.0.8.
```
Comment (SilverFire) [inline framework/web/User.php]: <code>Defaults to \`true\`</code>

File: framework/web/User.php [hunk 7]
```diff
@@ -669,6 +682,27 @@ public function can($permissionName, $params = [], $allowCaching = true)
     }
 
     /**
+     * Checks if the `Accept` header contains a content type that allows redirection to the login page.
+     * The login page is assumed to serve `text/html` or `application/xhtml+xml`, configure this via
+     * `$acceptableRedirectTypes`
```
Comment (SilverFire) [inline framework/web/User.php]: <code>\`$acceptableRedirectTypes\`</code> -> [[acceptableRedirectTypes]]

File: framework/web/User.php [hunk 8]
```diff
@@ -669,6 +682,27 @@ public function can($permissionName, $params = [], $allowCaching = true)
...
+        $acceptableTypes = Yii::$app->getRequest()->getAcceptableContentTypes();
+        if (empty($acceptableTypes)) {
+            return true;
+        }
+        foreach($acceptableTypes as $type => $params) {
+            if ($type === '*' || in_array($type, $this->acceptableRedirectTypes, true)) {
+                return true;
+            }
+        }
+        return false;
```
Comment (SilverFire) [inline framework/web/User.php]: Add a new line before return statement, please.

**Comment by SilverFire** [comment]
Merged: https://github.com/yiisoft/yii2/commit/040e102d97c7af909da829c02825cdd6ff923e0e

**Comment by SilverFire** [comment]
Thank you!

**Pull request closed without merging**